std/stdio.d

```diff
@@ -1522,6 +1524,143 @@ for every line.
+             the file an extension).
+    dir    = Directory of temporary file.
+  */
+    static File tempFile(string prefix = "", string suffix = "", string dir = "")
+    {
+        import std.exception : ErrnoException;
+        import std.file : isSymlink, tempDir;
+        import std.path : absolutePath, baseName, buildPath, dirName;
+
+        // For some reason, import std.file at the module level of std.stdio or
+        // even just inside of File results in a static assertion being hit in
+        // std.range.primitives.put. So, for now at least, we'll just assign
+        // dir here rather than using tempDir as a default argument so that the
+        // import can be local.
+        if(dir.empty)
+            dir = std.file.tempDir();
+
+        static string genTempName(string prefix, string suffix, string dir)
+        {
+            import std.ascii : digits, letters;
+            import std.random : rndGen;
+            rndGen.popFront();
+            auto name = new char[](prefix.length + 10 + rndGen.front % 5 + suffix.length);
+            name[0 .. prefix.length] = prefix;
+
+            char[letters.length + digits.length] chars;
+            chars[0 .. letters.length] = letters[];
+            chars[letters.length .. $] = digits[];
+
+            foreach(ref c; name[prefix.length .. $])
+            {
+                c = chars[rndGen.front % chars.length];
+                rndGen.popFront();
+            }
+
+            name[$ - suffix.length .. $] = suffix;
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        foreach(i; 0 .. 3)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // O_EXCL does not support symbolic links.
+            if(dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version(Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                enum flags = S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, flags);
+            }
+            else version(Windows)
+            {
+                enum flags = _S_IREAD | _S_IWRITE;
+                auto fd = _wsopen(tempCString!wchar(filename), O_RDWR | O_CREAT | O_EXCL, _SH_DENYNO, flags);
+            }
+            else static assert("Unsupported OS");
+
+            if(fd == -1)
+                continue;
+
+            File retval;
+            retval.fdopen(fd, "w+b", filename);
+            return retval;
+        }
+
+        version(Posix)        enum errorMsg = "Failed to create temporary file with open";
+        else version(Windows) enum errorMsg = "Failed to create temporary file with _wsopen";
```

Comment:
I was probably thinking that the way the files was opened was important for some reason, but I don't remember now, since it's been a while since I wrote that. But it probably would be better to not to put that information in the error message.